Visualize similar nuclei

mcovalt opened the pull request:
Allows the user to view similar nuclei. See demo video below.

https://user-images.githubusercontent.com/13244041/185230982-2cd83631-3f1b-483e-8570-582321db0f39.mov

mcovalt committed:
Make dedicated similar nuclei endpoint

This endpoint finds the closest nucleus and returns 5 similar nuclei.

mcovalt committed:
Add ability to clear all features of a layer

mcovalt committed:
Add map layer to show similar nuclei

mcovalt committed:
Update `MouseClickEvent` type

mcovalt committed:
Add sidebar to view similar nuclei

waxlamp commented on web/src/views/InvestigationDetail.vue:
```diff
@@ -486,8 +494,57 @@ export default defineComponent({
+    const similarNuclei: Ref<Array<DetectedStructureWithColor>> = ref([]);
+    const axios = inject('axios') as AxiosInstance;
+    function colorSimilarNuclei(nuclei: Array<DetectedStructure>) {
+      const colors = ['#FF0000', '#00FF00', '#0000FF', '#FFFF00', '#FF00FF', '#00FFFF'];
```
Don't use these colors. If you need categorical colors (which, see below), pull in D3's categorical colormaps, and use "category10".

Categorical colors to distinguish the markers may be the right way to go, but it could also be interesting to use a linear colormap to encode how similar each nucleus is. That might not be a good idea if you're always showing the 5 most similar (which presumably would all be close to the same color on a linear scale).

mcovalt replied:
Resolved in c865a1abb2f8639637eeece20cfedee49e408328. The new colors look much better!

waxlamp commented on web/src/views/InvestigationDetail.vue:
```diff
@@ -486,8 +494,57 @@ export default defineComponent({
+          return;
+        }
+        const url = `/similar-nuclei/${rootDataset.value.id}/${x}/${y}/`;
+        axios.get(url).then((resp) => {
```
Consider using async/await here.

mcovalt replied:
Awesome. More readable and less nesting. Resolved in a5609958c0259ed93e214de0e88b0f7ae2953355.

waxlamp commented on web/src/components/InvestigationSidebar.vue:
```diff
@@ -35,6 +35,12 @@
+        Similar Nuclei
+      </v-tab>
+      <v-tab-item>
+        <similar-nuclei-pane :similar-nuclei="similarNuclei" />
```
Use PascalCase for custom component declarations (https://v2.vuejs.org/v2/style-guide/?redirect=true#Component-name-casing-in-JS-JSX-strongly-recommended).

mcovalt replied:
Resolved in 5cc3fd867e4659042d5633a46ade59c4119ebdf8.

waxlamp commented on web/src/components/InvestigationSidebar.vue:
```diff
@@ -49,19 +55,34 @@
   },
 
-  setup() {
+  setup(props) {
```
Why this change?

mcovalt replied:
Resolved in 679e31ee02762064b10bd5e5febb71ede6fa683a.

waxlamp commented on web/src/components/SimilarNucleiPane.vue:
```diff
@@ -0,0 +1,53 @@
+      required: true,
+    },
+  },
+  setup(props) {
```
What do style guides recommend about including positional framework method arguments when you don't make use of them?

mcovalt replied:
Resolved in 679e31ee02762064b10bd5e5febb71ede6fa683a.

waxlamp reviewed (commented):
This LGTM (though I'm leaving my review as a comment so someone else has to provide a binding review) but I have a couple of ideas you could implement now or in a followup:
1. In addition to showing the 5 most similar nuclei, it might be interesting to also see the 5 *least* similar as well.
2. Change your "RGB" colors out for five colors from "Category10" (talk to me if you want some pairing on this).
3. Consider using a red-blue linear colormap to encode the dissimilar-to-similar range (especially in light of (1) above).
4. It would be great to also see an image of each nucleus in the search results.

Let me know what you think. Looks great!

annehaley reviewed (approved):
After Roni's suggestions are addressed, the rest looks good.

mcovalt committed:
Merge branch 'main' into visualize-similar-nuclei

mcovalt committed:
Use async/await

https://github.com/atlascope/atlascope/pull/208#discussion_r949158738

mcovalt committed:
Use PascalCase for custom component declarations

https://github.com/atlascope/atlascope/pull/208#discussion_r949162125

mcovalt committed:
Remove unused `props` argument

https://github.com/atlascope/atlascope/pull/208#discussion_r949162633
https://github.com/atlascope/atlascope/pull/208#discussion_r949164373

mcovalt committed:
Use `schemeCategory10` colors

https://github.com/atlascope/atlascope/pull/208#discussion_r949154864
